plugins/coverbrowser.koplugin/mosaicmenu.lua

```diff
@@ -0,0 +1,709 @@
+        end
+        -- Build new widgets
+        local texts_height = 0
+        if authors then
+            authors_wg = TextBoxWidget:new{
+                text = authors,
+                face = Font:getFace("cfont", math.max(self.authors_font_max - sizedec, self.authors_font_min)),
+                width = text_width,
+                alignment = "center",
+            }
+            texts_height = texts_height + authors_wg:getSize().h
+        end
+        if title then
+            title_wg = TextBoxWidget:new{
+                text = title,
+                face = Font:getFace("cfont", math.max(self.title_font_max - sizedec, self.title_font_min)),
+                width = text_width,
+                alignment = "center",
+            }
+            texts_height = texts_height + title_wg:getSize().h
+        end
+        if filename then
+            filename_wg = TextBoxWidget:new{
+                text = filename,
+                face = Font:getFace("cfont", math.max(self.filename_font_max - sizedec, self.filename_font_min)),
+                width = text_width,
+                alignment = "center",
+            }
+            texts_height = texts_height + filename_wg:getSize().h
+        end
+        local free_height = height - texts_height
+        if authors then
+            free_height = free_height - self.top_pad
+        end
+        if filename then
+            free_height = free_height - self.bottom_pad
+        end
+        inter_pad = math.floor(free_height / 2)
+
+        -- XXX We can benefit from adding to ui/widget/textboxwidget.lua at line 141
+        -- ("either a very long english word"):
+        --     if adjusted_idx == offset then self.has_split_inside_word = true end
+        -- The following as no effect till then
+        local textboxes_ok = true
+        if (authors_wg and authors_wg.has_split_inside_word) or (title_wg and title_wg.has_split_inside_word) then
+            -- We may get a nicer cover at next lower font size
+            textboxes_ok = false
+        end
+
+        if textboxes_ok and free_height > 0.2 * height then -- enough free space to not look constrained
+            break
+        end
+        -- (We may store the first widgets matching free space requirements but
+        -- not textboxes_ok, so that if we never ever get textboxes_ok candidate,
+        -- we can use them instead of the super-small strings-modified we'll have
+        -- at the end that are worse than the firsts)
+
+        sizedec = sizedec + self.sizedec_step
+        if sizedec > 20 then -- break out of loop when too small
+            -- but try a 2nd loop with some cleanup to strings (for filenames
+            -- with no space but hyphen or underscore instead)
+            if not loop2  then
+                loop2 = true
+                sizedec = self.initial_sizedec -- restart from initial big size
+                -- Replace underscores and hyphens with spaces, to allow text wrap there.
```

Comment:
Wait, we don't wrap on hyphens? Kind of sounds like that should probably be a thing. :-P